src/IO/H5/Dat.cpp

```diff
@@ -0,0 +1,220 @@
+ */
+template <typename T>
+auto vector_to_matrix(const std::vector<T>& raw_data,
+                      const std::array<hsize_t, 2>& size) {
+  std::vector<std::vector<T>> result(size[0]);
+  for (size_t l = 0; l < result.size(); ++l) {
+    result[l] = std::vector<T>(
+        raw_data.begin() + static_cast<std::ptrdiff_t>(l * size[1]),
+        raw_data.begin() + static_cast<std::ptrdiff_t>((l + 1) * size[1]));
+  }
+  return result;
+}
+}  // namespace
+
+namespace h5 {
+/// \cond HIDDEN_SYMBOLS
+Dat::Dat(const bool exists, detail::OpenGroup&& group, const hid_t location,
+         const std::string& name, std::vector<std::string> legend,
+         const uint32_t version)
+    : group_(std::move(group)),
+      name_(name + extension()),
+      version_(version),
+      legend_(std::move(legend)),
+      size_{{0, legend_.size()}} {
+  if (exists) {
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wold-style-cast"
+    dataset_id_ = H5Dopen2(location, name_.c_str(), H5P_DEFAULT);
+#pragma GCC diagnostic pop
+    CHECK_H5(dataset_id_, "Failed to open dataset");
+    {
+      const Version open_version(true, detail::OpenGroup{}, dataset_id_,
+                                 "version");
+      version_ = open_version.get_version();
+    }
+    {
+      const Header header(true, detail::OpenGroup{}, dataset_id_, "header");
+      header_ = header.get_header();
+    }
+
+    hid_t space_id = H5Dget_space(dataset_id_);
+    std::array<hsize_t, 2> maxdims{};
+    if (2 !=
+        H5Sget_simple_extent_dims(space_id, size_.data(), maxdims.data())) {
+      ERROR("Invalid number of dimensions in file on disk.");  // LCOV_EXCL_LINE
+    }
+    CHECK_H5(H5Sclose(space_id), "Failed to close dataspace");
+    legend_ = detail::read_strings_from_attribute(dataset_id_, "Legend"s);
+  } else {  // file does not exist
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wold-style-cast"
+    dataset_id_ = h5::detail::create_extensible_dataset(
+        location, name_, size_, std::array<hsize_t, 2>{{4, legend_.size()}},
+        {{H5S_UNLIMITED, legend_.size()}});
+#pragma GCC diagnostic pop
+    CHECK_H5(dataset_id_, "Failed to create dataset");
+
+    {
+      Version open_version(false, detail::OpenGroup{}, dataset_id_, "version",
+                           version_);
+    }
+    {
+      Header header(false, detail::OpenGroup{}, dataset_id_, "header");
+      header_ = header.get_header();
+    }
+    // Capitalized for compatibility with SpEC output
+    detail::write_strings_to_attribute(dataset_id_, "Legend"s, legend_);
+  }
+}
+
+Dat::~Dat() { CHECK_H5(H5Dclose(dataset_id_), "Failed to close dataset"); }
+/// \endcond HIDDEN_SYMBOLS
+
+void Dat::append(const std::vector<double>& data) {
+  if (data.size() != size_[1]) {
+    // We cannot test this error because the HDF5 file would not get properly
+    // closed and we will encounter the good old error:
+    // pure virtual method called terminate called recursively
```

Comment:
We get the error message from ERROR followed by the "pure virtual method" stuff several times. I'm not sure if this is HDF5 version/configuration dependent. In cases where this occurs in constructors we could throw inside the constructor and catch in `get` or `insert` function of the `File` class, then close the HDF5 file and ERROR. However, in functions like `append` this doesn't work so well. One thing I can try is passing the HDF5 file ID (basically the HDF5 equivalent of a file pointer) to the `h5::Object`s and only using it to close files in case of errors to avoid the virtual method vomit. What I'm not sure about is what would happen if a group is open when the `h5close` call occurs.

@kidder thoughts or ideas?